source/particle/world.cc

```diff
@@ -0,0 +1,889 @@
+    {
+      if (cell->active() && !cell->is_locally_owned())
+        return 0;
+
+      if (status == parallel::distributed::Triangulation<dim>::CELL_PERSIST
+          || status == parallel::distributed::Triangulation<dim>::CELL_REFINE)
+        {
+          const types::LevelInd found_cell = std::make_pair<int, int> (cell->level(),cell->index());
+          const unsigned int n_particles_in_cell = particles.count(found_cell);
+          return n_particles_in_cell * tracer_weight;
+        }
+      else if (status == parallel::distributed::Triangulation<dim>::CELL_COARSEN)
+        {
+          unsigned int n_particles_in_cell = 0;
+
+          for (unsigned int child_index = 0; child_index < cell->number_of_children(); ++child_index)
+            {
+              const typename parallel::distributed::Triangulation<dim>::cell_iterator child = cell->child(child_index);
+              const types::LevelInd found_cell = std::make_pair<int, int> (child->level(),child->index());
+              n_particles_in_cell += particles.count(found_cell);
+            }
+          return n_particles_in_cell * tracer_weight;
+        }
+
+      return 0;
+    }
+
+    template <int dim>
+    void
+    World<dim>::store_tracers(const typename parallel::distributed::Triangulation<dim>::cell_iterator &cell,
+                              const typename parallel::distributed::Triangulation<dim>::CellStatus status,
+                              void *data)
+    {
+      unsigned int n_particles_in_cell(0);
+
+      if (status == parallel::distributed::Triangulation<dim>::CELL_PERSIST
+          || status == parallel::distributed::Triangulation<dim>::CELL_REFINE)
+        {
+          const types::LevelInd found_cell = std::make_pair<int, int> (cell->level(),cell->index());
+          const std::pair<typename std::multimap<types::LevelInd, Particle<dim> >::iterator, typename std::multimap<types::LevelInd, Particle<dim> >::iterator> particles_in_cell
+            = particles.equal_range(found_cell);
+          n_particles_in_cell = std::distance(particles_in_cell.first,particles_in_cell.second);
+
+          unsigned int *ndata = static_cast<unsigned int *> (data);
+          *ndata++ = n_particles_in_cell;
+          data = static_cast<void *> (ndata);
+
+          for (typename std::multimap<types::LevelInd, Particle<dim> >::iterator particle = particles_in_cell.first;
+               particle != particles_in_cell.second; ++particle)
+            {
+              particle->second.write_data(data);
+            }
+          particles.erase(particles_in_cell.first,particles_in_cell.second);
+        }
+      else if (status == parallel::distributed::Triangulation<dim>::CELL_COARSEN)
+        {
+          const unsigned int coarsen_factor = (dim == 3) ? 8 : 4;
+
+          for (unsigned int child_index = 0; child_index < cell->number_of_children(); ++child_index)
```

Comment:
Do you really mean `number_of_children()` or would `n_children()` do? In other words, do you expect to refine/coarsen over more than one level at a time?